---
title: "scMEB Tutorial"
author: "Jiadi Zhu"
package: scMEB
date: "`r Sys.Date()`"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{scMEB Tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\VignetteDepends{SingleCellExperiment}
---


# Introduction
This package provides a novel and fast method for detecting single-cell DEGs 
without prior cell clustering results. The proposed method utilizes a small 
part of known non-DEGs (stably expressed genes) to build a minimum enclosing 
ball and defines the DEGs based on the distance of a mapped gene to the center 
of the hypersphere in a feature space. The method on this package is described 
in the article *scMEB: A fast and clustering-independent method for detecting 
differentially expressed genes in single-cell RNA-seq data* by Zhu, J.D and 
Yang, Y.L. (2023, pending publication).

# Preparations
To install the scMEB package into your R environment, start R and enter:
```{r, eval=FALSE}
library("devtools")
devtools::install_github("FocusPaka/scMEB")
```

Then, the scMEB package is ready to load.
```{r}
library(scMEB)
```

# The usage of the scMEB method
There is a example to introduce the usage of the function `scMEB()`:

1. Load the dependency package and the example scRNA-seq data
```{r, message = FALSE, warning = FALSE}
library(SingleCellExperiment)
```

The simulation data is generated by splatter package (Zappia L, et al. 2017).
The data include 5,000 genes and 100 cells.
```{r}
data(sim_scRNA_data)
sim_scRNA_data
```

We randomly sample 1,000 stable genes from the simulation data. 
```{r}
data(stable_gene)
head(stable_gene)
length(stable_gene)
```

2. Training a model for the simulation scRNA-seq data
```{r, message = FALSE, warning = FALSE}
sim_scRNA <- scMEB(sce=sim_scRNA_data, stable_idx=stable_gene, 
filtered = TRUE, gamma = seq(1e-04,0.001,1e-05), nu = 0.01, 
reject_rate = 0.1)
```

The parameters of the function `scMEB`:     
*   `sce` A SingleCellExperiment class scRNA-seq data. 

*   `stable_idx` A vector shows the name of stable expressed gene in sce. 

*   `filtered` A logical value to show if the data have been filtered.

*   `gamma` A parameter needed for all kernels except linear. 

*   `nu` A parameter needed for one-classification. 

*   `reject_rate` A value used in controling the scale of ball, default is 0.1. 

3. Predict a gene and find out whether DE gene

For *sim_data_sp* data, we predict the discrimination results as follows:
```{r}
sim_scRNA_pred <- predict(sim_scRNA$model, sim_scRNA$dat_pca)
summary(sim_scRNA_pred)
```
The discrimination result *TRUE*/*FALSE* correspond that gene is non-DE/DE 
gene.

scMEB also provides a metric for ranking the genes, that is, the distance 
between the gene and the sphere of the ball in the feature space. And the 
larger the distance is, the more likely it is that the gene is a DEG.
```{r}
table(sim_scRNA$dist>0)
```

```{r}
sim_scRNA_dist <- data.frame(Gene=rownames(sim_scRNA_data),
                             Distance=sim_scRNA$dist)
head(sim_scRNA_dist)
```

```{r}
sessionInfo()
```


